### PR TITLE
[CPU] Add missing BitCast lowering patterns.

### LINKED ISSUE #17780: iree-compile failures with n-D vector.bitcast ops
https://github.com/llvm/llvm-project/commit/137a7451f458cf7d8e1d88df93dbd8da6888886d enables support for n-D `arith.trunci` narrow type emulation. This causes iree-compile failures in the benchmark suite, so it is reverted in IREE by https://github.com/iree-org/iree/pull/17770.

The compiler failure can be reproduced with these steps:
1. reapply https://github.com/llvm/llvm-project/commit/137a7451f458cf7d8e1d88df93dbd8da6888886d in third_party/llvm-project
2. Compile the following IR with `iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=llvm-cpu --iree-input-type=none --iree-llvmcpu-target-triple=aarch64-none-linux-android29 --iree-opt-data-tiling=true --iree-llvmcpu-enable-ukernels=all --iree-llvmcpu-target-cpu-features=+dotprod --iree-vm-emit-polyglot-zip=true --iree-llvmcpu-debug-symbols=false linalg.mlir -o /tmp/linalg.vmfb`:
```mlir
module @jit_forward {
  util.func public @main(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub, iree.reflection = {iree.abi.declaration = "sync func @main(%input0: tensor<1x256xi8>) -> (%output0: tensor<1x2048xi32>)"}} {
    %cst = arith.constant dense_resource<__elided__> : tensor<256x2048xi4>
    %c0_i32 = arith.constant 0 : i32
    %0 = hal.tensor.import %arg0 "input0" : !hal.buffer_view -> tensor<1x256xi8>
    %collapsed = tensor.collapse_shape %0 [[0, 1]] : tensor<1x256xi8> into tensor<256xi8>
    %1 = tensor.empty() : tensor<2048xi32>
    %2 = linalg.fill ins(%c0_i32 : i32) outs(%1 : tensor<2048xi32>) -> tensor<2048xi32>
    %3 = linalg.vecmat ins(%collapsed, %cst : tensor<256xi8>, tensor<256x2048xi4>) outs(%2 : tensor<2048xi32>) -> tensor<2048xi32>
    %expanded = tensor.expand_shape %3 [[0, 1]] output_shape [1, 2048] : tensor<2048xi32> into tensor<1x2048xi32>
    %4 = hal.tensor.export %expanded "output0" : tensor<1x2048xi32> -> !hal.buffer_view
    util.return %4 : !hal.buffer_view
  }
}
```

This should cause the error:
```
error: LLVM Translation failed for operation: builtin.unrealized_conversion_cast
    %3 = linalg.vecmat ins(%collapsed, %cst : tensor<256xi8>, tensor<256x2048xi4>) outs(%2 : tensor<2048xi32>) -> tensor<2048xi32>
```

### PR DESCRIPTION
**Pull request by hanhanW**
The revision drops the local LLVM revert https://github.com/llvm/llvm-project/commit/137a7451f458cf7d8e1d88df93dbd8da6888886d

Fixes https://github.com/iree-org/iree/issues/17780

Co-authored-by: Andrzej Warzyński <andrzej.warzynski@arm.com>

### CODE REVIEW

**Comment by banach-space** [comment]
Thanks!

Mahesh suggested adding a test and I agree (https://github.com/iree-org/iree/pull/17790#issuecomment-2208104082):
> A test would be good to avoid regressions

Funnily, this seems to be already tested here:
* https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir#L61

From what I can tell, that broadcast gets linearised earlier on and hence why the tests passes. We should update it and this seems to work:
```diff
+++ b/compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir
@@ -54,7 +54,8 @@ module {
     %c4096 = arith.constant 4096 : index
     %c8192 = arith.constant 8192 : index
     %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c4096) flags(ReadOnly) : memref<128xi8, strided<[1], offset: 4096>>
-    %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c8192) : memref<256x64xi8, strided<[64, 1], offset: 8192>>
+    %out_buffer_i8 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c8192) : memref<256x64xi8, strided<[64, 1], offset: 8192>>
+    %out_buffer_i4 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%c8192) : memref<256x64xi4, strided<[64, 1], offset: 8192>>
     %2 = vector.load %0[%c0] : memref<128xi8, strided<[1], offset: 4096>>, vector<2xi8>
     %3 = vector.bitcast %2 : vector<2xi8> to vector<4xi4>
     %4 = vector.insert %3, %cst_0 [3] : vector<4xi4> into vector<4x4xi4>
@@ -67,10 +68,16 @@ module {
     %11 = vector.extract %9[1] : vector<4xi8> from vector<4x4xi8>
     %12 = vector.extract %9[2] : vector<4xi8> from vector<4x4xi8>
     %13 = vector.extract %9[3] : vector<4xi8> from vector<4x4xi8>
-    vector.store %10, %1[%c0, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
-    vector.store %11, %1[%c1, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
-    vector.store %12, %1[%c2, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
-    vector.store %13, %1[%c3, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
+
+    vector.store %10, %out_buffer_i8[%c0, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
+    vector.store %11, %out_buffer_i8[%c1, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
+    vector.store %12, %out_buffer_i8[%c2, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
+    vector.store %13, %out_buffer_i8[%c3, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
+
+    // Make sure that also n-D bitcast is lowered correctly
+    %14 = vector.bitcast %9 : vector<4x4xi8> to vector<4x8xi4>
+    vector.store %14, %out_buffer_i4[%c0, %c0] : memref<256x64xi4, strided<[64, 1], offset: 8192>>, vector<4x8xi4>
+
     return
   }
 }
diff --git a/third_party/llvm-project b/third_party/llvm-project
index 8ded6ce55d..1895ef314c 160000
```

I needs a bit of fine-tuning though (there's no need for two `vector.bitcast`s on 2D vectors).

Otherwise this LGTM, thanks!

**Comment by banach-space** [comment]
OK, I've got an updated version of `@interleave_and_bitcast_lowering` that should capture similar issues in the future (please feel free to include it in this PR @hanhanW ):
```mlir
module {
  func.func @interleave_and_bitcast_lowering() {
    %cst = arith.constant dense<4> : vector<4x2xi8>
    %cst_0 = arith.constant dense<0> : vector<4x4xi4>
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %c2 = arith.constant 2 : index
    %c3 = arith.constant 3 : index
    %c4096 = arith.constant 4096 : index
    %c8192 = arith.constant 8192 : index
    %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c4096) flags(ReadOnly) : memref<128xi8, strided<[1], offset: 4096>>
    %out_buffer = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%c8192) : memref<256x64xi4, strided<[64, 1], offset: 8192>>
    %2 = vector.load %0[%c0] : memref<128xi8, strided<[1], offset: 4096>>, vector<2xi8>
    %3 = vector.bitcast %2 : vector<2xi8> to vector<4xi4>
    %4 = vector.insert %3, %cst_0 [3] : vector<4xi4> into vector<4x4xi4>
    %5 = vector.bitcast %4 : vector<4x4xi4> to vector<4x2xi8>
    %6 = arith.shli %5, %cst : vector<4x2xi8>
    %7 = arith.shrsi %6, %cst : vector<4x2xi8>
    %8 = arith.shrsi %5, %cst : vector<4x2xi8>

    // Ops that should be lowered
    %9 = vector.interleave %7, %8 : vector<4x2xi8> -> vector<4x4xi8>
    %14 = vector.bitcast %9 : vector<4x4xi8> to vector<4x8xi4>

    vector.store %14, %out_buffer[%c0, %c0] : memref<256x64xi4, strided<[64, 1], offset: 8192>>, vector<4x8xi4>
    return
  }
}

// Make sure we can lower multi-dimensional `vector.interleave` and its
// corresponding multi-dimensional `vector.bitcast`.

// CHECK-LABEL: llvm.func @interleave_and_bitcast_lowering(
// vector.interleave should be gone entirely
//   CHECK-NOT:   vector.interleave
// 2D vector.bitcast tha followed should be replaced with 1D vector.bitcast
//       CHECK:   llvm.bitcast {{.*}} : vector<4xi8> to vector<8xi4>
//   CHECK-NOT:   vector.bitcast %{{.*}} : vector<4x4xi8> to vector<4x8xi4>
```

The current version of `@interleave_and_bitcast_lowering` is just insufficient - the compiler is clever enough to elide the bitcasts so there's nothing to "break" 😂

**Comment by hanhanW** [comment]
cool, nice! Thanks for the awesome update!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir [hunk 1]
```diff
@@ -54,23 +54,20 @@ module {
...
-    %11 = vector.extract %9[1] : vector<4xi8> from vector<4x4xi8>
-    %12 = vector.extract %9[2] : vector<4xi8> from vector<4x4xi8>
-    %13 = vector.extract %9[3] : vector<4xi8> from vector<4x4xi8>
-    vector.store %10, %1[%c0, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
-    vector.store %11, %1[%c1, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
-    vector.store %12, %1[%c2, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
-    vector.store %13, %1[%c3, %c0] : memref<256x64xi8, strided<[64, 1], offset: 8192>>, vector<4xi8>
+    %14 = vector.bitcast %9 : vector<4x4xi8> to vector<4x8xi4>
+
+    vector.store %14, %out_buffer[%c0, %c0] : memref<256x64xi4, strided<[64, 1], offset: 8192>>, vector<4x8xi4>
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir:70]: Do we want to replace this test or just add this as a different test?

**Review by qedawkins — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/convert_to_llvm.mlir:70]: I think the new one is more general and robust, so we don't need the old one anymore.

**Pull request merged**